app/src/main/java/org/mozilla/focus/permission/PermissionHandler.java

```diff
@@ -141,13 +146,18 @@ public void onRequestPermissionsResult(Context context, int requestCode, @NonNul
             return;
         }
         if (requestCode == actionId) {
-            setPermissionAsked(context, permissions[0]);
+
             if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                 permissionHandle.doActionGranted(permission, actionId, params);
                 clearAction();
             } else {
-                showSnackBar();
+                showToast();
+                // I'm not sure why it's so. This method already on Main thread.
```

Comment:
You need to do permissionNotGranted() here, note that we do call this in showSnackBar() but not in showToast(). Sorry for writing bad interfaces :P